rtCommon/bidsArchive.py

```diff
@@ -0,0 +1,571 @@
+        dataset root (/sub-01/ses-01) from being interpreted as being relative
+        to the root of the filesystem
+        """
+        if len(relPath) >= 1 and relPath[0] == "/":
+            return relPath[1:]
+        else:
+            return relPath
+
+    def absPathFromRelPath(self, relPath: str) -> str:
+        """
+        Makes an absolute path from the relative path within the dataset.
+        """
+        return os.path.join(self.data.root, self._stripRoot(relPath))
+
+    def fileExists(self, relPath: str) -> bool:
+        # search for matching BIDS file in the dataset
+        relPath = self._stripRoot(relPath)
+        result = self.data.get_file(self._stripRoot(relPath))
+        if result is not None:
+            return True
+
+    def dirExists(self, relPath: str) -> bool:
+        # search for matching directory in the dataset
+        return os.path.isdir(self.absPathFromRelPath(relPath))
+
+    def pathExists(self, path: str) -> bool:
+        return self.fileExists(path) or self.dirExists(path)
+
+    def findImage(self, path: str) -> nib.Nifti1Image:
+        """
+        Find an image within the dataset, if it exists.
+        """
+        # Validate extension and file existence
+        if isNiftiPath(path) and self.fileExists(path):
+            return readNifti(self.absPathFromRelPath(path))
+
+        return None
+
+    def _ensurePathExists(self, relPath: str):
+        """
+        Ensures the provided directory path exists in the dataset, creating
+        directories if needed.
+
+        Args:
+            relPath: Path to ensure existence of, relative to directory root.
+                All parts of the path are assumed to be directories.
+
+        Returns:
+            True if a layout update is needed, false otherwise.
+        """
+        if self.dirExists(relPath):
+            return False
+        else:
+            os.makedirs(self.absPathFromRelPath(relPath))
+            return True
+
+    def _updateLayout(self):
+        """
+        Updates the layout of the dataset so that any new metadata or image
+        files are added to the index.
+        """
+        # TODO(spolcyn): Find if there's a more efficient way to update the
+        # index that doesn't rely on implementation details of the PyBids (ie,
+        # the SQL DB it uses)
+        self.data = BIDSLayout(self.data.root)
+
+    def addImage(self, img: nib.Nifti1Image, path: str) -> None:
+        """
+        Add an image to the dataset at the provided path, creating the path if
+        it does not yet exist.
+        """
+        logger.debug("Writing new image to %s", self.absPathFromRelPath(path))
+        layoutUpdateRequired = self._ensurePathExists(os.path.dirname(path))
+
+        absPath = self.absPathFromRelPath(path)
+        nib.save(img, absPath)
+
+        if layoutUpdateRequired:
+            self._updateLayout()
+
+    def findMetadata(self, path: str) -> dict:
```

Comment:
Similar to above, but I'm not sure since it relates to many file types. Maybe getFileMetadata, or getComponentMetadata, ...